File: src/main/llm/openai.ts
import type { LlmSettings } from '@shared/types'
import { Provider, SiteRequest, systemPrompt, userPrompt, stripFences } from './provider'

interface OpenAiResponse {
  choices?: Array<{ message?: { content?: string } }>
}

export const openaiProvider: Provider = {
  async generateSite(request: SiteRequest, settings: LlmSettings): Promise<string> {
    const res = await fetch(`${settings.endpoint}/chat/completions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${settings.apiKey}`
      },
      body: JSON.stringify({
        model: settings.model,
        max_tokens: settings.maxTokens,
        messages: [
          { role: 'system', content: systemPrompt(settings.universe) },
          { role: 'user', content: userPrompt(request) }
        ]
      })
    })

    if (!res.ok) {
      throw new Error(`OpenAI request failed (${res.status}): ${await res.text()}`)
    }

    const data = (await res.json()) as OpenAiResponse
    const content = data.choices?.[0]?.message?.content
    if (!content) {
      throw new Error('OpenAI response contained no content')
    }
    return stripFences(content)
  }
}

File: crates/browser/src/llm/openai.rs
use serde::Deserialize;
use serde_json::json;

use super::GenerateError;
use crate::prompt::{strip_fences, system_prompt, user_prompt, SiteRequest};
use crate::settings::Settings;

const NAME: &str = "OpenAI";

#[derive(Deserialize)]
struct Response {
    #[serde(default)]
    choices: Vec<Choice>,
}

#[derive(Deserialize)]
struct Choice {
    message: Option<Message>,
}

#[derive(Deserialize)]
struct Message {
    content: Option<String>,
}

pub async fn generate(
    client: &reqwest::Client,
    settings: &Settings,
    request: &SiteRequest,
) -> Result<String, GenerateError> {
    let endpoint = settings.endpoint.trim_end_matches('/');
    let res = client
        .post(format!("{endpoint}/chat/completions"))
        .bearer_auth(&settings.api_key)
        .json(&json!({
            "model": settings.model,
            "max_tokens": settings.max_tokens,
            "messages": [
                { "role": "system", "content": system_prompt(&settings.universe) },
                { "role": "user", "content": user_prompt(request) }
            ]
        }))
        .send()
        .await
        .map_err(|e| GenerateError::Network(e.to_string()))?;

    let status = res.status();
    if !status.is_success() {
        let body = res.text().await.unwrap_or_default();
        return Err(GenerateError::Http {
            provider: NAME,
            status: status.as_u16(),
            body,
        });
    }
    let data: Response = res
        .json()
        .await
        .map_err(|e| GenerateError::Other(e.to_string()))?;
    let content = data
        .choices
        .into_iter()
        .next()
        .and_then(|c| c.message)
        .and_then(|m| m.content)
        .filter(|c| !c.is_empty())
        .ok_or(GenerateError::NoContent(NAME))?;
    Ok(strip_fences(&content))
}
